File: aoc15/ocaml/day_25/day_25.ml
let split = String.split_on_char ' ' (Input.read "input") in
let target_row = int_of_string (Input.num_drop_suffix (List.nth split 16)) in
let target_col = int_of_string (Input.num_drop_suffix (List.nth split 18)) in

let next (row, col, n) =
  if row = target_row && col = target_col then None
  else
    let row, col = if row = 1 then (col + 1, 1) else (row - 1, col + 1) in
    let n = n * 252533 mod 33554393 in
    Some (n, (row, col, n))
in

let first = Seq.fold_left (fun _ n -> n) 0 (Seq.unfold next (1, 1, 20151125)) in
Printf.printf "First: %d\n" first

File: aoc15/go/day_25/day_25.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	input, _ := os.ReadFile("input")
	split := strings.Fields(string(input))
	row, _ := strconv.Atoi(strings.TrimRight(split[15], ","))
	col, _ := strconv.Atoi(strings.TrimRight(split[17], "."))

	code := 20151125
	r := 1
	c := 1
	for {
		if r == 1 {
			r = c + 1
			c = 1
		} else {
			r--
			c++
		}
		code = (code * 252533) % 33554393
		if r == row && c == col {
			fmt.Println("First:", code)
			return
		}
	}
}
